src/mlpack/tests/kde_test.cpp

```diff
@@ -0,0 +1,821 @@
+  arma::vec bfEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  arma::vec treeEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  const double kernelBandwidth = 0.3;
+  const double relError = 0.01;
+
+  // Brute force KDE
+  GaussianKernel kernel(kernelBandwidth);
+  BruteForceKDE<GaussianKernel>(reference,
+                                query,
+                                bfEstimations,
+                                kernel);
+
+  // Optimized KDE
+  metric::EuclideanDistance metric;
+  KDE<GaussianKernel,
+      metric::EuclideanDistance,
+      arma::mat,
+      tree::Octree>
+    kde(relError, 0.0, kernel, KDEMode::DUAL_TREE_MODE, metric);
+  kde.Train(reference);
+  kde.Evaluate(query, treeEstimations);
+
+  // Check whether results are equal.
+  for (size_t i = 0; i < query.n_cols; ++i)
+    BOOST_REQUIRE_CLOSE(bfEstimations[i], treeEstimations[i], relError*100);
+}
+
+/**
+ * Test RTree dual-tree implementation results against brute force results.
+ */
+BOOST_AUTO_TEST_CASE(RTreeGaussianKDETest)
+{
+  arma::mat reference = arma::randu(2, 500);
+  arma::mat query = arma::randu(2, 200);
+  arma::vec bfEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  arma::vec treeEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  const double kernelBandwidth = 0.3;
+  const double relError = 0.01;
+
+  // Brute force KDE
+  GaussianKernel kernel(kernelBandwidth);
+  BruteForceKDE<GaussianKernel>(reference,
+                                query,
+                                bfEstimations,
+                                kernel);
+
+  // Optimized KDE
+  metric::EuclideanDistance metric;
+  KDE<GaussianKernel,
+      metric::EuclideanDistance,
+      arma::mat,
+      tree::RTree>
+    kde(relError, 0.0, kernel, KDEMode::DUAL_TREE_MODE, metric);
+  kde.Train(reference);
+  kde.Evaluate(query, treeEstimations);
+
+  // Check whether results are equal.
+  for (size_t i = 0; i < query.n_cols; ++i)
+    BOOST_REQUIRE_CLOSE(bfEstimations[i], treeEstimations[i], relError*100);
+}
+
+/**
+ * Test Standard Cover Tree dual-tree implementation results against brute
+ * force results.
+ */
+BOOST_AUTO_TEST_CASE(StandardCoverTreeGaussianKDETest)
+{
+  arma::mat reference = arma::randu(2, 500);
+  arma::mat query = arma::randu(2, 200);
+  arma::vec bfEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  arma::vec treeEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  const double kernelBandwidth = 0.3;
+  const double relError = 0.01;
+
+  // Brute force KDE
+  GaussianKernel kernel(kernelBandwidth);
+  BruteForceKDE<GaussianKernel>(reference,
+                                query,
+                                bfEstimations,
+                                kernel);
+
+  // Optimized KDE
+  metric::EuclideanDistance metric;
+  KDE<GaussianKernel,
+      metric::EuclideanDistance,
+      arma::mat,
+      tree::StandardCoverTree>
+    kde(relError, 0.0, kernel, KDEMode::DUAL_TREE_MODE, metric);
```

Comment:
Oops, this should be doubly indented.

Reply:
This is mostly fixed in 7bf036b0e87942fbbd9df7ef077fd31ccc25a6a9 :)